pvlib/pvsystem.py

```diff
@@ -421,6 +421,91 @@ def sapm_effective_irradiance(self, poa_direct, poa_diffuse,
             poa_direct, poa_diffuse, airmass_absolute, aoi,
             self.module_parameters, reference_irradiance=reference_irradiance)
 
+    def first_solar_spectral_correction(self, pw, airmass_absolute):
+        
+        """
+        Use the :py:func:`first_solar_spectral_correction` function to 
+        calculate the spectral loss modifier.
+
+        Parameters
+        ----------
+        pw : array-like
+            atmospheric precipitable water (cm).
+    
+        airmass_absolute : array-like
+            absolute (pressure corrected) airmass.
+
+        Returns
+        -------
+        modifier: array-like
+            spectral mismatch factor (unitless) which is can be multiplied
+            with broadband irradiance reaching a module's cells to estimate
+            effective irradiance, i.e., the irradiance that is converted to
+            electrical current.
+        """        
+
+        if self.module_parameters['fs_spectral_coefficients']:
+            coefficients = self.module_parameters['fs_spectral_coefficients']
+            module_type = None
+        else:
+            module_type = self._infer_cell_type()
+            coefficients = None
+
+        if (module_type is not None and coefficients is None) or \
```

Comment:
Recommend removing this if/else statement. If this method doesn't have the information necessary to successfully compute the spectral correction, then this method should say so rather than defaulting to 1. 

Consistent with Python's style, the ``atmosphere.first_solar_spectral_correction`` function throws a ``TypeError`` with some hint as to what the user needs to supply. We could (ideally/optionally) catch that error here and reraise it with a more specific message for how to modify ``module_parameters`` to make this work.

I am guessing that raising an exception here will cause issues in a ModelChain calculation for the few module/cell types for which ``_infer_cell_type`` mapping returns ``None``. One approach to avoid that may be to make the ``ModelChain.infer_spectral_model`` preemptively call ``pvsystem._infer_cell_type``.

Reply:
Good idea.  I'll move the test to ModelChain.infer_spectral_model and put a warning in place of that check in the actual functions.